gui/default/syncthing/core/pushNotificationsService.js

```diff
@@ -0,0 +1,86 @@
+angular.module('syncthing.core')
+    .service('PushNotifications', ['Events', function (Events) {
+        'use strict';
+
+        var self = this;
+
+        function checkNotificationPromise() {
+            try {
+                Notification.requestPermission().then();
+            } catch (e) {
+                return false;
+            }
+
+            return true;
+        }
+
+        function detectLocalStorage() {
+            // Feature detect localStorage; https://mathiasbynens.be/notes/localstorage-pattern
+            try {
+                var uid = new Date();
+                var storage = window.localStorage;
+                storage.setItem(uid, uid);
+                storage.removeItem(uid);
+                return storage;
+            } catch (exception) {
+                return undefined;
+            }
+        }
+
+        var _localStorage = detectLocalStorage();
+        var _SYNPN = "SYN_PN"; // const key for localStorage
+
+        angular.extend(self, {
+            isEnabled: function () {
+                return  self.isSupported() &&_localStorage && _localStorage[_SYNPN] === 'true'
```

Comment:
As of https://developers.google.com/web/fundamentals/push-notifications/permission-ux its recommended to ask user in our UI before triggering a native browser permission request.
Moreover, Chrome and Firefox will progressively silence permissions requests not triggered by an user interaction https://blog.chromium.org/2020/01/introducing-quieter-permission-ui-for.html

Personally, notifications are an intrusive feature and must activated only by the user himself if he wants/need it.

Notifications must be accepted site by site and browser by browser, to prevent notifications flooding. It’s how browsers work and we cannot do anything to that.

Reply:
> Looks like a good thing to build on, yet I suspect we'll have users screaming that they don't want this, or screaming that they want more (up to the levels of silly notification per file). Perhaps this could be configurable.
> 
> I've asked a few questions to see how we're taking this forward.

I agree with you. For the moment it is very basic to implement the core logic and concepts. Once this built more could be done to add more personalization per folders / per devices / per event types...

Reply:
Should we then store the choice in config rather than local storage? Whats the rationale for local storage?

Reply:
It is an assumption I've made that user not necessarily wants to enable push notifications on every devices.
I also wanted to reduce unwanted permissions prompts.

But as of I thinking about it it sounds better to add it to the config. If later we implement closed-tab notifications we need it in the config.

I'll push a commit to fix this today

Reply:
I think push notifications only work while the tab is open anyway, no?

Reply:
For the moment yes. But it can be implemented server-side to push notifications to closed tabs using Service Workers.

See https://github.com/SherClockHolmes/webpush-go

Reply:
Would that not be the desired approach to start with anyway?

It's not very useful to have notifications if you have to keep the tab open. You might as well look at the tab in that case?

How is this different from the service worker approach?

Reply:
Actual way: You need to keep tab open but not necessarily visible. Even if your browser isn't in the foreground notifications are shown.
Service worker: tab can be closed